storagenode/pieces/store.go

```diff
@@ -273,6 +274,78 @@ func (store *Store) Delete(ctx context.Context, satellite storj.NodeID, pieceID
 	return Error.Wrap(err)
 }
 
+// DeleteWithStorageFormat deletes the specified piece only at the specified format version
+func (store *Store) DeleteWithStorageFormat(ctx context.Context, satellite storj.NodeID, pieceID storj.PieceID, formatVer storage.FormatVersion) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	err = store.blobs.DeleteWithStorageFormat(ctx, storage.BlobRef{
+		Namespace: satellite.Bytes(),
+		Key:       pieceID.Bytes(),
+	}, formatVer)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	// delete records in both the piece_expirations and pieceinfo DBs, wherever we find it.
+	// both of these calls should return no error if the requested record is not found.
+	if store.expirationInfo != nil {
+		_, err = store.expirationInfo.DeleteExpiration(ctx, satellite, pieceID)
+	}
+	if store.v0PieceInfo != nil {
+		err = errs.Combine(err, store.v0PieceInfo.Delete(ctx, satellite, pieceID))
+	}
+	return Error.Wrap(err)
+
+}
+
+// MigrateV0ToV1 will migrate a piece stored with storage format v0 to storage
+// format v1. If the piece is not stored as a v0 piece it will return an error.
+// The follow failures are possible:
+// - Fail to open or read v0 piece. In this case no artifacts remain.
+// - Fail to Write or Commit v1 piece. In this case no artifacts remain.
+// - Fail to Delete v0 piece. In this case v0 piece may remain, but v1 piece
+//   will exist and be preferred in future calls.
+func (store *Store) MigrateV0ToV1(ctx context.Context, satelliteID storj.NodeID, pieceID storj.PieceID) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	info, err := store.v0PieceInfo.Get(ctx, satelliteID, pieceID)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	r, err := store.Reader(ctx, satelliteID, pieceID)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	w, err := store.Writer(ctx, satelliteID, pieceID)
+	if err != nil {
+		return Error.Wrap(errs.Combine(err, r.Close()))
+	}
+
+	_, err = io.Copy(w, r)
+	if err != nil {
+		return Error.Wrap(errs.Combine(err, r.Close(), w.Cancel(ctx)))
+	}
+
+	err = r.Close()
+	if err != nil {
+		return Error.Wrap(errs.Combine(err, w.Cancel(ctx)))
```

Comment:
If the reader close fails for some crazy reason, we still want to commit the writer and remove the v0 storage. We can and should return the error still, though

Reply:
I'm not explaining myself well. I'm suggesting using the deferred `r.Close()` call in my above comment in place of combining all return values with `r.Close()`, because of the high risk of somebody forgetting to add it at some point in the future. It looks like you've done everything right here, but it's a lot easier to check when there's a defer that covers all cases for you. Plus if the `r.Close()` is deferred, it can cleanly happen after `w.Commit()` and doesn't have to affect it.

Reply:
If the v0 reader is not closed, will it be able to delete the file? I can close it after the writer commit, but won't an open file descriptor prevent deleting the file?

Reply:
Dang, good point. Yeah, I don't think we can delete it on Windows without closing the reader first.

Reply:
See https://github.com/storj/storj/blob/master/storage/filestore/dir_windows.go#L131

Reply:
But, there still could be some other thing that's keeping the file from being deleted, e.g. anti-virus.

Reply:
Yeah, we have a try-again-later queue for deletions. Just good to avoid it when not necessary, because it's not persisted to disk.

Reply:
For the file locking case, rename should still work.

Reply:
On https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea , under `FILE_SHARE_DELETE`, it says "Note  Delete access allows both delete and rename operations". So it seems to imply that rename access isn't always allowed by default otherwise. Is that wrong? Or maybe "rename" isn't the same as "move" here?

Anyway perhaps we should add a monkit event for things being added to the delete-later queue, so we can see if it even happens at all in the wild.

Reply:
I know that the rename when you don't have file permissions to modify the file or the directories... but I'm not aware of any share specifically for renaming. And as far as I know by default renaming is always allowed... but yeah, I might be mis-remembering things.